### PR TITLE
Add a C++ version of the Virtual Keyboard example

### PR DESCRIPTION
**Pull request by tronical**

### CODE REVIEW

File: api/cpp/include/slint.h [hunk 1]
```diff
@@ -222,6 +223,11 @@ class WindowAdapterRc
         cbindgen_private::slint_windowrc_set_physical_size(&inner, &size);
     }
 
+    void dispatch_key_event(const slint::KeyEvent &event)
```
Comment (tronical) [inline api/cpp/include/slint.h:226]: On second thought, I'd like to replace this with two overloads that take the press and release event and remove the bare KeyEvent class.

**Review by ogoffart — DISMISSED** [review]

**Comment by ogoffart** [comment]
> On second thought, I'd like to replace this with two overloads that take the press and release event and remove the bare KeyEvent class.

Yet another way would be:

```cpp
void dispatch_key_press_event(const SharedString&);
void dispatch_key_release_event(const SharedString&);

// end then when we implement mouse:
void dispatch_pointer_press_event(Point, Button, ...);
void dispatch_pointer_mouse_event(Point, Button, ...);
//...
```

**Comment by ogoffart** [comment]
I'm not a big fan of the many function because on MCU, one may delegate the key processing to some keyboard controler, and one might want to return the event from that function.

Another option is to mimic rust API and have a WindowEvent

```c++
class WindowEvent {
  public: 
     static WindowEvent key_pressed(const SharedString&);
     static WindowEvent key_released(const SharedString&);
     
     static WindowEvent pointer_pressed(...);
     //...
   private:
     std::variant<cbingen_internal::KeyEvent, cbindgen_internal::PointerEvent> inner;
};
```

The problem i see is that there is no easy way to offer a getter API with this approach on the WindowEvent

File: api/cpp/include/slint_keyevent.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+public:
+private:
+    KeyEvent(const cbindgen_private::KeyInputEvent &event) : m_event(event) { }
+    cbindgen_private::KeyInputEvent m_event;
+    friend class private_api::WindowAdapterRc;
+    friend class KeyPressEvent;
+    friend class KeyReleaseEvent;
+};
+
+class KeyPressEvent : public KeyEvent
```
[the remark this answers is not in the document]
Comment (tronical) [reply api/cpp/include/slint_keyevent.h]: I went for the overload approach instead - that looks a little less awkward in the API IMO. We do have other key related input events such as composition, that we may want to stabilise in the future.

File: api/cpp/include/slint_platform.h [hunk 3]
```diff
@@ -148,7 +148,7 @@ class WindowAdapter : public AbstractWindowAdapter
     {
         private_api::assert_main_thread();
         if (was_initialized) {
-            cbindgen_private::slint_windowrc_dispatch_key_event(&self, event);
+            cbindgen_private::slint_windowrc_dispatch_key_event(&self, &event);
```
[the remark this answers is not in the document]
Comment (tronical) [reply api/cpp/include/slint_platform.h]: Agreed and removed.

File: examples/virtual_keyboard/cpp/main.cpp [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+#include "main_window.h"
+
+#include <ctime>
+
+void init_virtual_keyboard(slint::ComponentHandle<MainWindow> app)
+{
+    app->global<VirtualKeyboardHandler>().on_key_pressed(
+            [=](auto key) { app->window().dispatch_key_event(slint::KeyPressEvent(key)); });
```
[the remark this answers is not in the document]
Comment (tronical) [reply examples/virtual_keyboard/cpp/main.cpp]: Yeah, I was wondering the same after I pushed. I added a follow-up commit that fixes this in the Rust and C++ version.

**Comment by tronical** [comment]
> I'm not a big fan of the many function because on MCU, one may delegate the key processing to some keyboard controller, and one might want to return the event from that function.

I think that if we find a nice pattern for representing a big event enum, then we can introduce that as API later without any harm.

**Review by ogoffart — COMMENTED** [review]
What's the adventage of having

```cpp
app->window().dispatch_key_event(slint::KeyPressEvent(key));
```

over

```cpp
app->window().dispatch_key_press_event(key);
```

I.e: why do we add the KeyPressEvent struct. It's only ever going to be used with this one function anyway.

One advantage i see is that one could return a `std::variant<slint::KeyPressEvent, slint::KeyReleaseEvent>` and use 
```cpp
std::variant the_event = get_keyboard_event();
std::visit([&](const auto &e) { app->window().dispatch_key_event(e); }, the_event);
```

But i'm not sure if this is really a valid use case.

**Comment by tronical** [comment]
> What's the adventage of having
> 
> ```c++
> app->window().dispatch_key_event(slint::KeyPressEvent(key));
> ```
> 
> over
> 
> ```c++
> app->window().dispatch_key_press_event(key);
> ```
> 
> I.e: why do we add the KeyPressEvent struct. It's only ever going to be used with this one function anyway.

Ahh, yes, I like that! I'll go with that.

File: examples/virtual_keyboard/cpp/main.cpp [hunk 5]
```diff
@@ -0,0 +1,21 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+#include "main_window.h"
+
+#include <ctime>
```
[the remark this answers is not in the document]
Comment (tronical) [reply examples/virtual_keyboard/cpp/main.cpp:6]: Me, too. clangd somehow added it ;-). I'll remove it.

File: examples/virtual_keyboard/cpp/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+cmake_minimum_required(VERSION 3.21)
+project(slint_cpp_virtual_keyboard LANGUAGES CXX)
+
+if (NOT TARGET Slint::Slint)
+    find_package(Slint REQUIRED)
+endif()
+
+add_executable(virtual_keyboard main.cpp)
+if (MSVC)
+    target_compile_options(virtual_keyboard PRIVATE /bigobj)
```
[the remark this answers is not in the document]
Comment (tronical) [reply examples/virtual_keyboard/cpp/CMakeLists.txt:13]: We could do that perhaps yes. Not quite in scope of this PR, but I, too, wonder why we didn't do it like that before. Maybe I can find something in the mattermost history.

File: api/cpp/include/slint_platform.h [hunk 7]
```diff
@@ -142,16 +142,6 @@ class WindowAdapter : public AbstractWindowAdapter
         }
     }
 
-    /// Send a key event to this window
-    // Note: in rust, this is on the Window. FIXME: use a public event type
-    void dispatch_key_event(const cbindgen_private::KeyInputEvent &event)
-    {
-        private_api::assert_main_thread();
```
[the remark this answers is not in the document]
Comment (tronical) [reply api/cpp/include/slint_platform.h:149]: Good point!

File: examples/virtual_keyboard/cpp/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (tronical) [reply examples/virtual_keyboard/cpp/CMakeLists.txt:13]: https://github.com/slint-ui/slint/pull/2551

File: CHANGELOG.md [hunk 8]
```diff
@@ -18,6 +18,8 @@ All notable changes to this project are documented in this file.
 
 - macOS: Fixed `install_name` for `libslint_cpp.dylib` use `@rpath` instead of absolute paths to the build directory.
 - Fixed memory corruption when embedding images in generated C++ code.
+- Add support for dispatching key press and key release events to `slint::Window` via `dispatch_key_event`. This removes
+  the `dispatch_key_event` function in `slint::experimental::platform::WindowAdapter::dispatch_key_event`.
```
Comment (ogoffart) [inline CHANGELOG.md:22]:
```suggestion
- Add support for dispatching key press and key release events to `slint::Window` via `dispatch_key_*_event`. 
  This replaces the experimental `slint::experimental::platform::WindowAdapter::dispatch_key_event`.
```

**Review by ogoffart — APPROVED** [review]

File: CHANGELOG.md (hunk 8 above)
Comment (tronical) [reply CHANGELOG.md:22]: Thanks! Will apply.

**Pull request merged**